### PR TITLE
Implements support for external package loading in validator.

### PR DESCRIPTION
**Pull request by dom96**
Moves package loading to `prepareWasmLinearMemory` so that `preloadDynamicLibs` can use the loaded packages instead of the embedded ones.

Also fixes a stack-use-after-scope error in `getPyodidePackage`

### Test Plan

```
$ bazel test @workerd//src/cloudflare/internal/test/... --config=asan
```

### CODE REVIEW

File: src/workerd/api/pyodide/pyodide.c++ [hunk 1]
```diff
@@ -35,10 +35,9 @@ void PyodideBundleManager::setPyodideBundleData(
       kj::mv(version), {.messageReader = kj::mv(messageReader), .bundle = bundle});
 }
 
-const kj::Maybe<kj::ArrayPtr<const unsigned char>> PyodidePackageManager::getPyodidePackage(
+const kj::Maybe<const kj::Array<unsigned char>&> PyodidePackageManager::getPyodidePackage(
```
Comment (hoodmane) [inline src/workerd/api/pyodide/pyodide.c++:38]: What's the reason for this change?

File: src/pyodide/python-entrypoint-helper.ts [hunk 2]
```diff
@@ -176,11 +163,12 @@ try {
   if (!IS_TRACING) {
     if (IS_WORKERD) {
       // If we're in workerd, we have to do setupPackages in the IoContext, so don't start it yet.
+      //
+      // Also, not doing this here seems to cause @workerd//src/cloudflare/internal/test/d1:python-d1-api-test
+      // to fail under --config=asan.
+      //
       // TODO: fix this.
       await getPyodide();
```
Comment (dom96) [inline src/pyodide/python-entrypoint-helper.ts:171]: If I remove this call then some of our workerd tests fail under asan.

https://paste.cfdata.org/GdKtTZYLiz6LktnWP5gP

Hood mentioned this may be related to the isolate pool changes. @danlapid any ideas what might be going on here?

File: src/workerd/api/pyodide/pyodide.c++ (hunk 1 above)
Comment (dom96) [reply src/workerd/api/pyodide/pyodide.c++:38]: I was getting stack-use-after-scope errors before this change. I'm guessing it was returning a pointer to the reference, rather than the array itself. This seems to fix it.

File: src/pyodide/internal/python.ts [hunk 3]
```diff
@@ -62,6 +64,11 @@ async function prepareWasmLinearMemory(Module: Module): Promise<void> {
   mountSitePackages(Module, SITE_PACKAGES.rootInfo);
   entropyMountFiles(Module);
   Module.noInitialRun = !SHOULD_RESTORE_SNAPSHOT;
+
+  // NB. loadPackages adds the packages to the `SITE_PACKAGES` global which then gets used in
+  // preloadDynamicLibs.
+  await loadPackages(Module, TRANSITIVE_REQUIREMENTS);
```
Comment (hoodmane) [inline src/pyodide/internal/python.ts:70]: I wonder line 70 and lines 64 and 65 should be moved into `loadPyodide`. Since they don't strictly speaking have anything to do with `prepareWasmLinearMemory`. And is it worth giving `loadPackages` its own jaeger span?

File: src/pyodide/python-entrypoint-helper.ts [hunk 4]
```diff
@@ -119,7 +105,8 @@ function getMainModule(): Promise<PyModule> {
     }
     mainModulePromise = (async function () {
       const pyodide = await getPyodide();
-      await setupPackages(pyodide);
+      patchLoadPackage(pyodide);
```
Comment (hoodmane) [inline src/pyodide/python-entrypoint-helper.ts:108]: Maybe move this back into `setupPatches`? It sounds related.

File: src/pyodide/internal/setupPackages.ts [hunk 5]
```diff
@@ -44,6 +45,7 @@ class SitePackagesDir {
       path: '',
       name: '',
       parts: [],
+      reader: null,
```
Comment (hoodmane) [inline src/pyodide/internal/setupPackages.ts:48]: What about setting `reader: EmbeddedPackagesTarReader` and removing the `node.reader ?? EmbeddedPackagesTarReader` logic in `snapshot.ts`? Would that work?

**Review by hoodmane — APPROVED** [review]
I made a few minor questions and suggestions.

File: src/pyodide/internal/setupPackages.ts (hunk 5 above)
Comment (dom96) [reply src/pyodide/internal/setupPackages.ts:48]: Probably, but doesn't matter much. I plan to get rid of `EmbeddedPackagesTarReader` in a follow up PR.

File: src/pyodide/python-entrypoint-helper.ts (hunk 2 above)
Comment (dom96) [reply src/pyodide/python-entrypoint-helper.ts]: Had to remove it again. No way around it. I'll probably disable these tests under asan for now but we should fix this asap, this asan issue is likely lurking there right now and just isn't being triggered.

File: .bazelrc [hunk 6]
```diff
@@ -188,6 +188,7 @@ build:asan --config=sanitizer-common
 build:asan --copt="-fsanitize=address" --linkopt="-fsanitize=address"
 build:asan --test_env=ASAN_OPTIONS=abort_on_error=true
 build:asan --test_env=KJ_CLEAN_SHUTDOWN=1
+build:asan --test_tag_filters=-no-asan
```
[the remark this answers is not in the document]
Comment (dom96) [reply .bazelrc:191]: good catch, thanks!

File: src/pyodide/python-entrypoint-helper.ts (hunk 2 above)
Comment (danlapid) [reply src/pyodide/python-entrypoint-helper.ts]: Pool code hasn't been merged yet so I don't see how it might be related
